src/jltypes.c

```diff
@@ -810,6 +815,27 @@ jl_value_t *jl_cache_type_(jl_datatype_t *type)
     return (jl_value_t*)type;
 }
 
+int jl_type_equality_is_identity(jl_value_t *t1, jl_value_t *t2)
+{
+    if (t1 == t2)
+        return 1;
+    if (!jl_is_datatype(t1) || !jl_is_datatype(t2))
+        return 0;
+    jl_datatype_t *dt1 = (jl_datatype_t *) t1;
+    jl_datatype_t *dt2 = (jl_datatype_t *) t2;
+    if (!(is_cacheable(dt1) || jl_svec_len(dt1->parameters) == 0)) {
+        if (!(is_cacheable(dt2) || jl_svec_len(dt2->parameters) == 0))
+            return 0;
+        else
+            return 1;
+    }
+    if (!(is_cacheable(dt2) || jl_svec_len(dt2->parameters) == 0))
+        return 1;
+
+    return is_typekey_ordered(jl_svec_data(dt1->parameters), jl_svec_len(dt1->parameters)) ==
+        is_typekey_ordered(jl_svec_data(dt2->parameters), jl_svec_len(dt2->parameters));
```

Comment:
This predicate seems pretty expensive. If we're going to use this, we should probably cache this property like we do for has-free-typevars.

Reply:
Sure. Performance-wise, I'm more worried about taking the fast-paths in subtype.c much less often, though. BTW, does nanosoldier have any benchmarks that exercise the type system?

> If we're going to use this

Before I set into motion working on the caching: What are the chances of that becoming true? (And what are the alternatives?)